Upcoming breaking changes in Numpy 2.0

NumPy 2.0 hasn't been released yet, but the [initial release notes](https://github.com/numpy/numpy/releases/tag/v2.0.0b1) from the pre-release mentions some breaking changes that they address in their [migration guide](https://numpy.org/devdocs/numpy_2_0_migration_guide.html). I don't know how the Cantera team wants to handle compatibility going forward, but I thought I would go ahead and make a thread in case it's helpful.

# Previous Cantera builds

I installed NumPy 2.0.0b1 and attempted to import Cantera 3.0.0 which yields the following error:

```
...
    from ._cantera import *
  File "cantera/_cantera.pyx", line 33, in init cantera._cantera
  File "cantera/_utils.pyx", line 1, in init cantera._utils
ValueError: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject
```

Clearly previous builds of Cantera can only be used with NumPy 1.x and will not work with 2.x releases. I don't know if there is a way to enforce `numpy<2.0.0` as a package requirement on PyPI, or if that can't be done retroactively, but it might need to be added to some documentation somewhere.

# Future Cantera builds

I attempted to build Cantera from source from the latest commit using NumPy 2.0.0b1 to see what changes would need to be made. The first build error that I get from the breaking changes is below:

```
build\python\cantera\_onedim.cpp(8544): error C2039: 'subarray': is not a member of '_PyArray_Descr'
...
build\python\cantera\_onedim.cpp(8545): error C2039: 'subarray': is not a member of '_PyArray_Descr'
```

The changes to the `PyArray_Descr` struct are covered in [this section](https://numpy.org/devdocs/numpy_2_0_migration_guide.html#the-pyarray-descr-struct-has-been-changed) of the migration guide. It seems the `PyDataType_SUBARRAY` accessor function needs to be used. They also provide a `npy2_compat.h` file that can be included in the code base that will allow compilation on NumPy 1.x and not affect 2.x builds.

I haven't gotten any further yet, but I will edit this if I do.

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -450,8 +450,10 @@ jobs:
           find samples/python -type f -iname "*.py" \
           -exec sh -c 'for n; do echo "$n" | tee -a results.txt && python3 "$n" >> results.txt || exit 1; done' sh {} +
         env:
-          # The ignore setting here is due to a new warning introduced in Matplotlib==3.6.0
-          PYTHONWARNINGS: "error,ignore:warn_name_set_on_empty_Forward::pyparsing,ignore:datetime.datetime.utcfromtimestamp:DeprecationWarning:"
+          # The pyparsing ignore setting is due to a new warning introduced in Matplotlib==3.6.0
+          # @todo: Remove the trapz-related ignore when dropping support for NumPy 1.x
+          #     and replacing np.trapz with np.trapezoid
+          PYTHONWARNINGS: "error,ignore:warn_name_set_on_empty_Forward::pyparsing,ignore:datetime.datetime.utcfromtimestamp:DeprecationWarning:,ignore:`trapz`:DeprecationWarning"
           MPLBACKEND: Agg
       - name: Save the results file for inspection
         uses: actions/upload-artifact@v4

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -39,8 +39,8 @@ jobs:
       - name: Install Apt dependencies
         run: |
           sudo apt update
-          sudo apt install libboost-dev gfortran libopenmpi-dev libpython3-dev \
-          libblas-dev liblapack-dev libhdf5-dev
+          sudo apt install libboost-dev gfortran libopenmpi-dev libblas-dev \
+            liblapack-dev libhdf5-dev
           gcc --version
       - name: Upgrade pip
         run: python3 -m pip install -U pip setuptools wheel
@@ -59,13 +59,20 @@ jobs:
           select-by-folder: /home/runner/work/cantera/cantera/test/matlab_experimental
 
   ubuntu-multiple-pythons:
-    name: ${{ matrix.os }} with Python ${{ matrix.python-version }}
+    name: ${{ matrix.os }} with Python ${{ matrix.python-version }}, Numpy ${{ matrix.numpy || 'latest' }}
     runs-on: ${{ matrix.os }}
     timeout-minutes: 60
     strategy:
       matrix:
         python-version: ['3.8', '3.10', '3.11', '3.12']
         os: ['ubuntu-20.04', 'ubuntu-22.04']
+        numpy: ['']
+        include:
+          # Keep some test cases with NumPy 1.x until we drop support
+        - python-version: '3.12'
+          os: 'ubuntu-24.04'
+          numpy: "'<2.0'"
+
       fail-fast: false
     env:
       HDF5_LIBDIR: /usr/lib/x86_64-linux-gnu/hdf5/serial
@@ -83,15 +90,16 @@ jobs:
     - name: Install Apt dependencies
       run: |
         sudo apt update
-        sudo apt install libboost-dev gfortran libopenmpi-dev libpython3-dev \
-        libblas-dev liblapack-dev libhdf5-dev libfmt-dev
+        sudo apt install libboost-dev gfortran libopenmpi-dev libblas-dev \
+          liblapack-dev libhdf5-dev libfmt-dev
         gcc --version
     - name: Upgrade pip
       run: python3 -m pip install -U pip setuptools wheel
     - name: Install Python dependencies
       run: |
-        python3 -m pip install ruamel.yaml scons==3.1.2 numpy cython pandas pytest \
-        pytest-github-actions-annotate-failures pytest-xdist pint graphviz
+        python3 -m pip install ruamel.yaml scons==3.1.2 numpy${{ matrix.numpy }} \
+          cython pandas pytest pytest-github-actions-annotate-failures pytest-xdist \
+          pint graphviz
     - name: Build Cantera
       run: |
         python3 `which scons` build env_vars=all -j4 debug=n --debug=time \
@@ -393,14 +401,38 @@ jobs:
             "${DOCS_OUTPUT_DIR}" ${RSYNC_USER}@${RSYNC_SERVER}:${RSYNC_DEST}
 
   run-examples:
-    name: Run Python ${{ matrix.python-version }} examples on ${{ matrix.os }}
+    name: Run Python ${{ matrix.python-version }} examples on ${{ matrix.os }}, NumPy ${{ matrix.numpy || 'latest' }}
     runs-on: ${{ matrix.os }}
     timeout-minutes: 60
     needs: ["ubuntu-multiple-pythons"]
     strategy:
       matrix:
-        os: ["ubuntu-20.04", "ubuntu-22.04"]
-        python-version: ['3.8', '3.10', '3.11', '3.12']
+        # Keep some test cases with NumPy 1.x until we explicitly drop support
+        include:
+        - os: "ubuntu-20.04"
+          python-version: "3.8"
+          numpy: ""
+        - os: "ubuntu-22.04"
+          python-version: "3.8"
+          numpy: ""
+        - os: "ubuntu-20.04"
+          python-version: "3.10"
+          numpy: "==1.21.6"
+        - os: "ubuntu-22.04"
+          python-version: "3.10"
+          numpy: ""
+        - os: "ubuntu-22.04"
+          python-version: "3.11"
+          numpy: "==1.23.5"
+        - os: "ubuntu-20.04"
+          python-version: "3.11"
+          numpy: ""
+        - os: "ubuntu-22.04"
+          python-version: "3.12"
+          numpy: "==1.26.4"
+        - os: "ubuntu-20.04"
+          python-version: "3.12"
+          numpy: ""
       fail-fast: false
     env:
       HDF5_LIBDIR: /usr/lib/x86_64-linux-gnu/hdf5/serial
@@ -433,8 +465,8 @@ jobs:
         run: python3 -m pip install -U pip setuptools wheel
       - name: Install Python dependencies
         run: |
-            python3 -m pip install numpy ruamel.yaml pandas pyarrow matplotlib scipy \
-              pint graphviz CoolProp
+            python3 -m pip install numpy${{ matrix.numpy }} ruamel.yaml pandas pyarrow \
+              matplotlib scipy pint graphviz CoolProp
             python3 -m pip install --pre --no-index --find-links dist cantera
       - name: Run the examples
         # See https://unix.stackexchange.com/a/392973 for an explanation of the -exec part.
@@ -460,7 +492,7 @@ jobs:
         with:
           path: results.txt
           retention-days: 2
-          name: example-results-${{ matrix.python-version }}-${{ matrix.os }}
+          name: example-results-${{ matrix.python-version }}-${{ matrix.os }}-np${{ matrix.numpy || 'latest' }}
         # Run this step if the job was successful or failed, but not if it was cancelled
         # Using always() would run this step if the job was cancelled as well.
         if: failure() || success()

--- a/samples/python/onedim/flame_speed_convergence_analysis.py
+++ b/samples/python/onedim/flame_speed_convergence_analysis.py
@@ -305,10 +305,10 @@ def analyze_errors(grids, speeds, true_speed):
     and compare these with our estimated errors.
     This will show if our estimates are reasonable, or conservative, or too optimistic.
     """
-    true_speed_estimates = np.full_like(speeds, np.NaN)
-    total_percent_error_estimates = np.full_like(speeds, np.NaN)
-    actual_extrapolated_percent_errors = np.full_like(speeds, np.NaN)
-    actual_raw_percent_errors = np.full_like(speeds, np.NaN)
+    true_speed_estimates = np.full_like(speeds, np.nan)
+    total_percent_error_estimates = np.full_like(speeds, np.nan)
+    actual_extrapolated_percent_errors = np.full_like(speeds, np.nan)
+    actual_raw_percent_errors = np.full_like(speeds, np.nan)
     for i in range(3, len(grids)):
         print(grids[: i + 1])
         true_speed_estimate, total_percent_error_estimate = extrapolate_uncertainty(

--- a/site_scons/buildutils.py
+++ b/site_scons/buildutils.py
@@ -972,9 +972,9 @@ def compare_profiles(
         "pos. err"
     ))
     header.append(f"{10*'-'}  -----  {14*'-'}  {14*'-'}  {9*'-'}  {9*'-'}  {9*'-'}")
-    ref_ptp = reference.ptp(axis=1)
+    ref_ptp = np.ptp(reference, axis=1)
     ref_max = np.abs(reference).max(axis=1)
-    sample_ptp = sample.ptp(axis=1)
+    sample_ptp = np.ptp(sample, axis=1)
     sample_max = np.abs(sample).max(axis=1)
     scale = np.maximum(
         np.maximum(ref_ptp[1:], ref_max[1:]),

--- a/test/python/test_onedim.py
+++ b/test/python/test_onedim.py
@@ -487,11 +487,11 @@ class TestFreeFlame(utilities.CanteraTest):
         self.sim.transport_model = 'unity-Lewis-number'
         self.sim.set_refine_criteria(ratio=3.0, slope=0.08, curve=0.12)
         self.sim.solve(loglevel=0, auto=True)
-        dh_unity_lewis = self.sim.enthalpy_mass.ptp()
+        dh_unity_lewis = np.ptp(self.sim.enthalpy_mass)
 
         self.sim.transport_model = 'mixture-averaged'
         self.sim.solve(loglevel=0)
-        dh_mix = self.sim.enthalpy_mass.ptp()
+        dh_mix = np.ptp(self.sim.enthalpy_mass)
 
         # deviation of enthalpy should be much lower for unity Le model (tends
         # towards zero as grid is refined)

--- a/test/python/test_reaction.py
+++ b/test/python/test_reaction.py
@@ -1851,7 +1851,7 @@ class TestExtensible3(utilities.CanteraTest):
 class InterfaceReactionTests(ReactionTests):
     # test suite for surface reaction expressions
 
-    _value = np.NAN # reference value
+    _value = np.nan # reference value
     _coverage_deps = None
 
     @classmethod

--- a/test/python/test_reactor.py
+++ b/test/python/test_reactor.py
@@ -2383,7 +2383,9 @@ class TestReactorSensitivities(utilities.CanteraTest):
         return dtdp
 
     # See https://github.com/Cantera/enhancements/issues/55
+    # @todo: replace np.trapz with np.trapezoid when dropping support for NumPy 1.x
     @unittest.skip("Integration of sensitivity ODEs is unreliable")
+    @pytest.mark.filterwarnings("ignore:`trapz` is deprecated")
     def test_ignition_delay_sensitivity(self):
         species = ('H2', 'H', 'O2', 'H2O2', 'H2O', 'OH', 'HO2')
         dtigdh_cvodes = self.calc_dtdh(species)

--- a/test/python/test_thermo.py
+++ b/test/python/test_thermo.py
@@ -1208,6 +1208,8 @@ class TestPlasmaPhase(utilities.CanteraTest):
                                (ct.avogadro * ct.electron_charge))
         self.assertNear(mean_electron_energy, self.phase.mean_electron_energy)
 
+    # @todo: replace np.trapz with np.trapezoid when dropping support for NumPy 1.x
+    @pytest.mark.filterwarnings("ignore:`trapz` is deprecated")
     def test_discretized_electron_energy_distribution(self):
         levels = np.array([0.0, 1.0, 10.0])
         dist = np.array([0.0, 0.9, 0.01])

--- a/test/python/utilities.py
+++ b/test/python/utilities.py
@@ -181,10 +181,10 @@ def compareProfiles(reference, sample, rtol=1e-5, atol=1e-12, xtol=1e-5):
     bad = []
     template = '{0:9.4e}  {1: 3d}   {2:14.7e}  {3:14.7e}  {4:9.3e}  {5:9.3e}  {6:9.3e}'
     for i in range(1, nVars):
-        scale = max(max(abs(reference[i])), reference[i].ptp(),
-                    max(abs(sample[i])), sample[i].ptp())
+        scale = max(max(abs(reference[i])), np.ptp(reference[i]),
+                    max(abs(sample[i])), np.ptp(sample[i]))
         slope = np.zeros(nTimes)
-        slope[1:] = np.diff(reference[i]) / np.diff(reference[0]) * reference[0].ptp()
+        slope[1:] = np.diff(reference[i]) / np.diff(reference[0]) * np.ptp(reference[0])
 
         comp = np.interp(reference[0], sample[0], sample[i])
         for j in range(nTimes):

